src/plugins/data/public/search/expressions/esaggs.ts

```diff
@@ -0,0 +1,90 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { get } from 'lodash';
+import { StartServicesAccessor } from 'src/core/public';
+import { Adapters } from 'src/plugins/inspector/common';
+import {
+  EsaggsExpressionFunctionDefinition,
+  EsaggsStartDependencies,
+  getEsaggsMeta,
+  handleEsaggsRequest,
+} from '../../../common/search/expressions';
+import { DataPublicPluginStart, DataStartDependencies } from '../../types';
+
+function createEsaggs({
+  getStartDependencies,
+}: {
+  getStartDependencies: () => Promise<EsaggsStartDependencies>;
+}) {
+  return (): EsaggsExpressionFunctionDefinition => ({
+    ...getEsaggsMeta(),
+    async fn(input, args, { inspectorAdapters, abortSignal, getSearchSessionId }) {
+      const {
+        addFilters,
+        aggs,
+        deserializeFieldFormat,
+        indexPatterns,
+        searchSource,
+      } = await getStartDependencies();
+
+      const aggConfigsState = JSON.parse(args.aggConfigs);
+      const indexPattern = await indexPatterns.get(args.index);
+      const aggConfigs = aggs.createAggConfigs(indexPattern, aggConfigsState);
+
+      return await handleEsaggsRequest(input, args, {
+        abortSignal: (abortSignal as unknown) as AbortSignal,
+        addFilters,
+        aggs: aggConfigs,
+        deserializeFieldFormat,
+        filters: get(input, 'filters', undefined),
+        indexPattern,
+        inspectorAdapters: inspectorAdapters as Adapters,
+        metricsAtAllLevels: args.metricsAtAllLevels,
+        partialRows: args.partialRows,
+        query: get(input, 'query', undefined) as any,
+        searchSessionId: getSearchSessionId(),
+        searchSourceService: searchSource,
+        timeFields: args.timeFields,
+        timeRange: get(input, 'timeRange', undefined),
+      });
+    },
+  });
+}
+
+/** @internal */
+export function getEsaggs({
+  getStartServices,
+}: {
+  getStartServices: StartServicesAccessor<DataStartDependencies, DataPublicPluginStart>;
+}) {
+  return createEsaggs({
+    getStartDependencies: async () => {
+      const [, , self] = await getStartServices();
+      const { fieldFormats, indexPatterns, query, search } = self;
+      return {
+        addFilters: query.filterManager.addFilters.bind(query.filterManager),
+        aggs: search.aggs,
+        deserializeFieldFormat: fieldFormats.deserialize.bind(fieldFormats),
+        indexPatterns,
+        searchSource: search.searchSource,
+      };
+    },
+  });
```

Comment:
This is the main difference between the `public` and `server` implementations -- basically it's a function that wires up all of the stateful dependencies on various `start` contracts.

The `createEsaggs` function above is much simpler: it's just responsible for retrieving the needed pieces of state from those services, and calling the helper that handles the entire request.

The goal here was to isolate differences between client & server code so that we could still share as much as possible from `common`, and I've used a similar approach with `indexPatternLoad`. Open to any feedback or other ideas on this.

Reply:
Discussing this further [in this thread](https://github.com/elastic/kibana/pull/84590/files/66da84d2265243282213f3d13b12d0dfb28383ed..4e7cf167bcf6a1f11fc7bedaaed589cc4ffe3859#r534258195).